utils/pattern/src/single.rs

```diff
@@ -0,0 +1,256 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Code for the [`SinglePlaceholder`] pattern backend.
+
+use core::{cmp::Ordering, str::FromStr};
+use writeable::Writeable;
+
+use crate::common::*;
+use crate::Error;
+
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+
+/// A singleton enum for the [`SinglePlaceholder`] pattern backend.
+///
+/// # Examples
+///
+/// ```
+/// use core::cmp::Ordering;
+/// use icu_pattern::PatternItem;
+/// use icu_pattern::SinglePlaceholder;
+/// use icu_pattern::SinglePlaceholderKey;
+/// use icu_pattern::SinglePlaceholderPattern;
+///
+/// // Parse the string syntax and check the resulting data store:
+/// let pattern =
+///     SinglePlaceholderPattern::try_from_str("Hello, {0}!").unwrap();
+///
+/// assert_eq!(
+///     pattern.iter().cmp(
+///         [
+///             PatternItem::Literal("Hello, "),
+///             PatternItem::Placeholder(SinglePlaceholderKey::Singleton),
+///             PatternItem::Literal("!")
+///         ]
+///         .into_iter()
+///     ),
+///     Ordering::Equal
+/// );
+/// ```
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[allow(clippy::exhaustive_enums)] // Singleton
+pub enum SinglePlaceholderKey {
+    Singleton,
+}
+
+impl FromStr for SinglePlaceholderKey {
```

Comment:
do we want to enforce that it's `{0}`? Seems like a typo waiting to happen that we could catch.

Reply:
Happy to have others weigh in here. I wrote this to accept any arbitrary placeholder with the constraint that there is at most one placeholder in the pattern string. I see no harm in accepting `"Hello, {name}!"`-type patterns.